### ISSUE TITLE
Approx Epsilon Usage & Constructors

### ISSUE THREAD

**Issue by phil-zxx** [onqtam/doctest#389]
Hi,

As I often deal with floating point values, I tend to use the `Approx` class quite often. Now, let's consider the below example. The `CHECK` will pass even though the true absolute difference is `0.00001 = 1.e-05`. If we use `Approx(testValue).epsilon(1.e-12)` then the test will give `CHECK(false)`.
```cpp
const double testValue = 0.98765;
const double trueValue = 0.98766;
CHECK(Approx(testValue) == trueValue);  // this gives CHECK(true)
```
Usually, I deem values to be correct if they pass a certain `absolute` threshold:
```cpp
abs(lhs - rhs) < epsilon;                                                   // (1)
```
However, `doctest` implements this formula:
```cpp
abs(lhs - rhs) < epsilon * ( rhs.m_scale + max(abs(lhs), abs(rhs)) );       // (2)
```
which is slightly different.

---

Questions:

1. Why does `epsilon` default to `std::numeric_limits<float>::epsilon()*100` = `1.2e-05` as opposed to `std::numeric_limits<double>::epsilon()*100` = `2.2e-14`. I am asking since all comparisons are based on `double` values, and not `float` values. Personally, I find `1.2e-05` to be quite a low threshold.<br>

2. In `catch2` one can write `Approx(testValue).margin(1.e-12).epsilon(0)` to exactly achieve method `(1)` above.
(a) Is there benefit in adding extra settings to doctest's `Approx` class? So that one can achieve both method `(1)` and `(2)`?
(b) In order to avoid these long factory-like expressions (as `catch2` has them), one could allow to add them via an additional constructor. E.g. `Approx(double value, double epsilon, double margin)` would help shorten the expression to `Approx(testValue, 1.e-12, 0)`.

Happy to help write the code for this, but wanted to get some feedback on my questions first.

**Comment by phil-zxx** [comment]
Just thought I'd post the approx class that I use on my side (often instead of the native `Approx`). If this `ApproxAbs` class would ship with `doctest` it would be really useful.

```cpp
struct ApproxAbs
{
    ApproxAbs(double value, double tolerance = 1.e-12)
        : m_value(value), m_tolerance(tolerance) { }

    bool operator==(const double& rhs) const {
        return std::abs(this->m_value - rhs) < m_tolerance;
    }

    bool operator!=(const double& rhs) const {
        return !(*this == rhs);
    }

    friend std::ostream& operator<<(std::ostream& os, const ApproxAbs& rhs) {
        const auto digits = static_cast<std::streamsize>(1 + std::log10(1. / rhs.m_tolerance));
        return os << "ApproxAbs(" << std::setprecision(digits) << rhs.m_value << ")";
    }

private:
    const double m_value, m_tolerance;
};
```

**Comment by anders-wind** [comment]
I don't agree your wanted style changes - I like the builder/fluent pattern and it would even be possible to not affect performance of anyone by using that pattern. If `Approx(my_val).abs_epsilon(my_eps)` returns a new type `ApproxRelAndAbs`(or whatever) then only those who need it pay for the additional checks.

Having library support for absolute difference would however be nice!

**Comment by cdeln** [comment]
Ok, there are two discussions in this thread
1) Issues with current Approx implementation
2) Suggestion of adding an absolute difference checker

Regarding 1)
I agree that defaulting to `std::numeric_limits<float>::epsilon()` is not ideal when working with higher precision floating point types (`double` and `long double`). In C++ this could be mitigated by making turning `Approx` into a template class and use [class template argument deduction](https://en.cppreference.com/w/cpp/language/class_template_argument_deduction) to make this transparent. Unfortunately we can not do this without breaking downstream consumers who rely on the current default value of epsilon, so that would be a 3.xx discussion (we can still implement a "new Approx" which can have new default behavior).

Regarding 2)
Note that this is a near duplicate to https://github.com/doctest/doctest/issues/360 .
I am also in favor of the idea of adding more classes for different types of floating point comparisons.
I recently also added this feature suggestion to check floating in ulps, which is also a form of absolute difference test: https://github.com/doctest/doctest/issues/864

We could experiment with these features in a "floating point extension" which potentially could be incorporated into the core of a later version of doctest.
